### DISCUSSION TITLE
SDFat example bench.ino does not compile

### DISCUSSION THREAD

**Discussion by Dresch123** [earlephilhower/arduino-pico#296] [Q&A]
Hi,
SDFat library example "bench.ino" compiles and runs with Arduino Pico Mbed but with this core I get the error below. Could someone point out what I am doing wrong? Thanks, Bill
bench:68:1: error: 'SdFat' does not name a type
   68 | SdFat sd;
      | ^~~~~
bench:69:1: error: 'File' does not name a type
   69 | File file;
      | ^~~~
bench:84:1: error: 'ArduinoOutStream' does not name a type; did you mean 'ArduinoStream_h'?
   84 | ArduinoOutStream cout(Serial);
      | ^~~~~~~~~~~~~~~~
      | ArduinoStream_h
C:\Users\HOME\AppData\Local\Temp\arduino_modified_sketch_994260\bench.ino: In function 'void cidDmp()':
bench:90:3: error: 'cid_t' was not declared in this scope; did you mean 'sdfat::cid_t'?
   90 |   cid_t cid;
      |   ^~~~~
      |   sdfat::cid_t
In file included from C:\Users\HOME\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\1.8.2\libraries\SdFat\src/SdCard/SdCardInterface.h:28,
                 from C:\Users\HOME\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\1.8.2\libraries\SdFat\src/SdCard/SdioCard.h:28,
                 from C:\Users\HOME\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\1.8.2\libraries\SdFat\src/SdCard/SdCard.h:27,
                 from C:\Users\HOME\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\1.8.2\libraries\SdFat\src/SdFat.h:32,
                 from C:\Users\HOME\AppData\Local\Temp\arduino_modified_sketch_994260\bench.ino:4:
C:\Users\HOME\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\1.8.2\libraries\SdFat\src/SdCard/SdCardInfo.h:297:27: note: 'sdfat::cid_t' declared here
  297 | } __attribute__((packed)) cid_t;
      |                           ^~~~~
bench:91:8: error: 'sd' was not declared in this scope; did you mean 'std'?
   91 |   if (!sd.card()->readCID(&cid)) {
      |        ^~
      |        std
bench:91:28: error: 'cid' was not declared in this scope
   91 |   if (!sd.card()->readCID(&cid)) {
      |                            ^~~
bench:95:3: error: 'cout' was not declared in this scope
   95 |   cout << F("\nManufacturer ID: ");
      |   ^~~~
bench:96:11: error: 'hex' was not declared in this scope; did you mean 'sdfat::hex'?
   96 |   cout << hex << int(cid.mid) << dec << endl;
      |           ^~~
      |           sdfat::hex
In file included from C:\Users\HOME\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\1.8.2\libraries\SdFat\src/iostream/istream.h:31,
                 from C:\Users\HOME\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\1.8.2\libraries\SdFat\src/iostream/iostream.h:31,
                 from C:\Users\HOME\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\1.8.2\libraries\SdFat\src/iostream/bufstream.h:32,
                 from C:\Users\HOME\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\1.8.2\libraries\SdFat\src/iostream/ArduinoStream.h:32,
                 from C:\Users\HOME\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\1.8.2\libraries\SdFat\src/sdios.h:31,
                 from C:\Users\HOME\AppData\Local\Temp\arduino_modified_sketch_994260\bench.ino:5:
C:\Users\HOME\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\1.8.2\libraries\SdFat\src/iostream/ios.h:265:18: note: 'sdfat::hex' declared here
  265 | inline ios_base& hex(ios_base& str) {
      |                  ^~~
bench:96:22: error: 'cid' was not declared in this scope
   96 |   cout << hex << int(cid.mid) << dec << endl;
      |                      ^~~
bench:96:34: error: 'dec' was not declared in this scope; did you mean 'sdfat::dec'?
   96 |   cout << hex << int(cid.mid) << dec << endl;
      |                                  ^~~
      |                                  sdfat::dec
In file included from C:\Users\HOME\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\1.8.2\libraries\SdFat\src/iostream/istream.h:31,
                 from C:\Users\HOME\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\1.8.2\libraries\SdFat\src/iostream/iostream.h:31,
                 from C:\Users\HOME\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\1.8.2\libraries\SdFat\src/iostream/bufstream.h:32,
                 from C:\Users\HOME\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\1.8.2\libraries\SdFat\src/iostream/ArduinoStream.h:32,
                 from C:\Users\HOME\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\1.8.2\libraries\SdFat\src/sdios.h:31,
                 from C:\Users\HOME\AppData\Local\Temp\arduino_modified_sketch_994260\bench.ino:5:
C:\Users\HOME\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\1.8.2\libraries\SdFat\src/iostream/ios.h:257:18: note: 'sdfat::dec' declared here
  257 | inline ios_base& dec(ios_base& str) {
      |                  ^~~
bench:96:41: error: 'endl' was not declared in this scope; did you mean 'sdfat::endl'?
   96 |   cout << hex << int(cid.mid) << dec << endl;
      |                                         ^~~~
      |                                         sdfat::endl
In file included from C:\Users\HOME\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\1.8.2\libraries\SdFat\src/iostream/bufstream.h:32,
                 from C:\Users\HOME\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\1.8.2\libraries\SdFat\src/iostream/ArduinoStream.h:32,
                 from C:\Users\HOME\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\1.8.2\libraries\SdFat\src/sdios.h:31,
                 from C:\Users\HOME\AppData\Local\Temp\arduino_modified_sketch_994260\bench.ino:5:
C:\Users\HOME\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\1.8.2\libraries\SdFat\src/iostream/iostream.h:50:17: note: 'sdfat::endl' declared here
   50 | inline ostream& endl(ostream& os) {
      |                 ^~~~
C:\Users\HOME\AppData\Local\Temp\arduino_modified_sketch_994260\bench.ino: In function 'void setup()':
bench:125:5: error: 'SysCall' has not been declared
  125 |     SysCall::yield();
      |     ^~~~~~~
bench:128:3: error: 'cout' was not declared in this scope
  128 |   cout << F("\nUse a freshly formatted SD for best performance.\n");
      |   ^~~~
bench:135:11: error: 'uppercase' was not declared in this scope; did you mean 'sdfat::uppercase'?
  135 |   cout << uppercase << showbase << endl;
      |           ^~~~~~~~~
      |           sdfat::uppercase
In file included from C:\Users\HOME\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\1.8.2\libraries\SdFat\src/iostream/istream.h:31,
                 from C:\Users\HOME\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\1.8.2\libraries\SdFat\src/iostream/iostream.h:31,
                 from C:\Users\HOME\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\1.8.2\libraries\SdFat\src/iostream/bufstream.h:32,
                 from C:\Users\HOME\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\1.8.2\libraries\SdFat\src/iostream/ArduinoStream.h:32,
                 from C:\Users\HOME\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\1.8.2\libraries\SdFat\src/sdios.h:31,
                 from C:\Users\HOME\AppData\Local\Temp\arduino_modified_sketch_994260\bench.ino:5:
C:\Users\HOME\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\1.8.2\libraries\SdFat\src/iostream/ios.h:385:18: note: 'sdfat::uppercase' declared here
  385 | inline ios_base& uppercase(ios_base& str) {
      |                  ^~~~~~~~~
bench:135:24: error: 'showbase' was not declared in this scope; did you mean 'sdfat::showbase'?
  135 |   cout << uppercase << showbase << endl;
      |                        ^~~~~~~~
      |                        sdfat::showbase
In file included from C:\Users\HOME\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\1.8.2\libraries\SdFat\src/iostream/istream.h:31,
                 from C:\Users\HOME\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\1.8.2\libraries\SdFat\src/iostream/iostream.h:31,
                 from C:\Users\HOME\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\1.8.2\libraries\SdFat\src/iostream/bufstream.h:32,
                 from C:\Users\HOME\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\1.8.2\libraries\SdFat\src/iostream/ArduinoStream.h:32,
                 from C:\Users\HOME\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\1.8.2\libraries\SdFat\src/sdios.h:31,
                 from C:\Users\HOME\AppData\Local\Temp\arduino_modified_sketch_994260\bench.ino:5:
C:\Users\HOME\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\1.8.2\libraries\SdFat\src/iostream/ios.h:353:18: note: 'sdfat::showbase' declared here
  353 | inline ios_base& showbase(ios_base& str) {
      |                  ^~~~~~~~
bench:135:36: error: 'endl' was not declared in this scope; did you mean 'sdfat::endl'?
  135 |   cout << uppercase << showbase << endl;
      |                                    ^~~~
      |                                    sdfat::endl
In file included from C:\Users\HOME\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\1.8.2\libraries\SdFat\src/iostream/bufstream.h:32,
                 from C:\Users\HOME\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\1.8.2\libraries\SdFat\src/iostream/ArduinoStream.h:32,
                 from C:\Users\HOME\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\1.8.2\libraries\SdFat\src/sdios.h:31,
                 from C:\Users\HOME\AppData\Local\Temp\arduino_modified_sketch_994260\bench.ino:5:
C:\Users\HOME\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\1.8.2\libraries\SdFat\src/iostream/iostream.h:50:17: note: 'sdfat::endl' declared here
   50 | inline ostream& endl(ostream& os) {
      |                 ^~~~
C:\Users\HOME\AppData\Local\Temp\arduino_modified_sketch_994260\bench.ino: In function 'void loop()':
bench:150:3: error: 'cout' was not declared in this scope
  150 |   cout << F("Type any character to start\n");
      |   ^~~~
bench:152:5: error: 'SysCall' has not been declared
  152 |     SysCall::yield();
      |     ^~~~~~~
bench:155:46: error: 'endl' was not declared in this scope; did you mean 'sdfat::endl'?
  155 |   cout << F("FreeStack: ") << FreeStack() << endl;
      |                                              ^~~~
      |                                              sdfat::endl
In file included from C:\Users\HOME\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\1.8.2\libraries\SdFat\src/iostream/bufstream.h:32,
                 from C:\Users\HOME\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\1.8.2\libraries\SdFat\src/iostream/ArduinoStream.h:32,
                 from C:\Users\HOME\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\1.8.2\libraries\SdFat\src/sdios.h:31,
                 from C:\Users\HOME\AppData\Local\Temp\arduino_modified_sketch_994260\bench.ino:5:
C:\Users\HOME\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\1.8.2\libraries\SdFat\src/iostream/iostream.h:50:17: note: 'sdfat::endl' declared here
   50 | inline ostream& endl(ostream& os) {
      |                 ^~~~
bench:161:8: error: 'sd' was not declared in this scope; did you mean 's'?
  161 |   if (!sd.begin(SD_CONFIG)) {
      |        ^~
      |        s
bench:34:19: error: 'SdSpiConfig' was not declared in this scope; did you mean 'sdfat::SdSpiConfig'?
   34 | #define SD_CONFIG SdSpiConfig(SD_CS_PIN, DEDICATED_SPI, SPI_CLOCK)
      |                   ^~~~~~~~~~~
C:\Users\HOME\AppData\Local\Temp\arduino_modified_sketch_994260\bench.ino:161:17: note: in expansion of macro 'SD_CONFIG'
  161 |   if (!sd.begin(SD_CONFIG)) {
      |                 ^~~~~~~~~
In file included from C:\Users\HOME\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\1.8.2\libraries\SdFat\src/SdCard/SdSpiCard.h:35,
                 from C:\Users\HOME\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\1.8.2\libraries\SdFat\src/SdCard/SdCard.h:28,
                 from C:\Users\HOME\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\1.8.2\libraries\SdFat\src/SdFat.h:32,
                 from C:\Users\HOME\AppData\Local\Temp\arduino_modified_sketch_994260\bench.ino:4:
c:\users\home\appdata\local\arduino15\packages\rp2040\hardware\rp2040\1.8.2\libraries\sdfat\src\spidriver\sdspidriver.h:145:7: note: 'sdfat::SdSpiConfig' declared here
  145 | class SdSpiConfig {
      |       ^~~~~~~~~~~
bench:164:7: error: 'sd' was not declared in this scope; did you mean 's'?
  164 |   if (sd.fatType() == FAT_TYPE_EXFAT) {
      |       ^~
      |       s
bench:164:23: error: 'FAT_TYPE_EXFAT' was not declared in this scope; did you mean 'sdfat::FAT_TYPE_EXFAT'?
  164 |   if (sd.fatType() == FAT_TYPE_EXFAT) {
      |                       ^~~~~~~~~~~~~~
      |                       sdfat::FAT_TYPE_EXFAT
In file included from C:\Users\HOME\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\1.8.2\libraries\SdFat\src/ExFatLib/ExFatVolume.h:27,
                 from C:\Users\HOME\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\1.8.2\libraries\SdFat\src/ExFatLib/ExFatLib.h:27,
                 from C:\Users\HOME\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\1.8.2\libraries\SdFat\src/SdFat.h:33,
                 from C:\Users\HOME\AppData\Local\Temp\arduino_modified_sketch_994260\bench.ino:4:
C:\Users\HOME\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\1.8.2\libraries\SdFat\src/ExFatLib/ExFatPartition.h:41:15: note: 'sdfat::FAT_TYPE_EXFAT' declared here
   41 | const uint8_t FAT_TYPE_EXFAT = 64;
      |               ^~~~~~~~~~~~~~
bench:170:31: error: 'sd' was not declared in this scope; did you mean 's'?
  170 |   cout << F("Card size: ") << sd.card()->sectorCount()*512E-9;
      |                               ^~
      |                               s
bench:176:8: error: 'file' was not declared in this scope
  176 |   if (!file.open("bench.dat", O_RDWR | O_CREAT | O_TRUNC)) {
      |        ^~~~
bench:199:5: error: 'file' was not declared in this scope
  199 |     file.truncate(0);
      |     ^~~~
bench:242:5: error: 'file' was not declared in this scope
  242 |     file.rewind();
      |     ^~~~
bench:278:3: error: 'file' was not declared in this scope
  278 |   file.close();
      |   ^~~~
Multiple libraries were found for "SdFat.h"
 Used: C:\Users\HOME\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\1.8.2\libraries\SdFat
 Not used: C:\Users\HOME\Documents\Arduino\libraries\SdFat
exit status 1
'SdFat' does not name a type

**Comment by earlephilhower** [comment]
That example is in the `examples.donotuse` for a reason. :)  The problem is the base SdFat from grieman(sp?) conflicts with the full C++ libs and the Arduino `File` class that's implemented here and on the ESP8266.  So, please don't use it as the conflicting `File`s and other issues will cause build failures as you see.

Look at the `LittleFS` `SpeedTest example.  You can change a `#define` and use SDFat instead of LittleFS (documented in the INO) and get some raw #s quickly.

**Comment by Dresch123** [comment]
Many thanks Earle. I will give that a try this evening. I had not looked at the SDFat directory within your core directory. I had just downloaded and installed SDFat thru the Library Manager. 
If I can ask one more question: When using Bodmer;s TFT_sSPI with your core versus the Mbed core (for example Bodmer's bouncing circles INO), the Mbed core runs ~3X faster: 45fps for the Mbed and 17fps for your core. Am I doing something wrong or are there ways I can speed that up?
https://github.com/Bodmer/TFT_eSPI/commit/637f7f1fe36b9040d893f0a72f117233617397fd
I really prefer the flexibility of your core (especially the multicore techniques) but am looking for a bit of raw speed for a project I am working on.
Thanks again,
Bill

**Reply by earlephilhower** [reply]
Don't use the upstream SdFat library or you'll end up with the errors you saw in pretty much any code that does SD File accesses (due to the File redefinition he uses).  You may need to uninstall the library manager version of SdFat or may end up with strange problems where the IDE will take some files from the external SdFat lib and some from the included core SdFat lab.

For the SPI stuff, I really don't know.  As it's a 1-man operation here I just leveraged the SDK's SPI APIs and went for easy and correct.  They're 100% sync and very basic, whereas I believe the MBED code wrote an SPI driver from scratch for the HW that supports pipelining/etc.

**Comment by Dresch123** [comment]
Earle, please don't take the question about the frame rate as a dig. I am in awe of your accomplishment with the Pico. A massive undertaking and truly impressive results. THANK YOU VERY MUCH. 
And thanks for the advice concerning the SDFat. I will try that. Bill

**Reply by earlephilhower** [reply]
No worries, it wasn't taken in a bad way!  If you do come up with a PR to speed up SPI, as always it would be much appreciated!

**Reply by Dresch123** [reply]
Hi Earle,
I have been seeing a number of Pico Pi interface boards with the micro SD interface wired for SDIO/SD MMC. 
https://www.adafruit.com/product/5160
and 
https://www.waveshare.com/product/raspberry-pi/boards-kits/pico-eval-board.htm

I know that the Arduino code for the ESP32 supports 1 and 4 bit SD MMC using SD_MMC.h 
When I have tested the ESP32 code, the SDIO 4 bit moved data 3 to 4 times faster than the SPI interface.

Will these libraries work within the Pico Pic Arduino code?
Sorry if this is a stupid question!
Bill

**Comment by Dresch123** [comment]
https://github.com/raspberrypi/pico-extras/tree/master/src/rp2_common/pico_sd_card

"1 and 4 bit SDIO support using PIO. 
This is functional (currently writing is only 1 bit), but the the code is very much prototype and the API is just a placeholder - 
the command set needs to be separated from the SDIO and shared with SPI"

So it appears there is prototype code to do this at the official github site.

**Reply by earlephilhower** [reply]
Unfortunately the issue is that, as it says, it is "very much prototype".  :(

Assuming that it did work and an Arduino-compatible API was implemented, then you'd also need to update the grieman's `SdFat` and the downstream version used here to call the new functions.

It's definitely within the realm or possibility, but it's probably not a small task.
